src/tester.rs

```diff
@@ -569,30 +583,103 @@ fn run_cmd(
+    let mut buf = [0; READBUF];
+    let start = Instant::now();
+    let mut last_warning = Instant::now();
+    let mut next_warning = last_warning
+        .checked_add(Duration::from_secs(TIMEOUT))
+        .unwrap();
     loop {
-        match child.wait_timeout(Duration::from_secs(TIMEOUT)).unwrap() {
-            Some(_) => break,
-            None => {
-                if inner.test_threads == 1 {
-                    eprint!("running for over {} seconds... ", TIMEOUT * looped);
-                } else {
-                    eprintln!(
-                        "\nlang_tests::{} ... has been running for over {} seconds",
-                        test_fname,
-                        TIMEOUT * looped
-                    );
-                }
+        let timeout = {
+            // FIXME: When Rust 1.39.0 is out we can replace this mess with duration_since_checked.
+            let t = Instant::now();
+            if t > next_warning {
+                Duration::from_secs(1)
+            } else {
+                next_warning.duration_since(t)
             }
         };
-        looped += 1;
+        if let Ok(_) = poll(&mut pollfds, Some(timeout)) {
+            if pollfds[0].revents & POLLIN == POLLIN {
+                while let Ok(i) = stderr.read(&mut buf) {
+                    if i == 0 {
+                        break;
+                    }
+                    cap_stderr.extend_from_slice(&buf[..i]);
+                    if inner.nocapture {
+                        eprint!("{}", str::from_utf8(&buf).unwrap());
+                    }
+                }
+            }
+
+            if pollfds[0].revents & POLLIN == POLLIN {
+                while let Ok(i) = stdout.read(&mut buf) {
+                    if i == 0 {
+                        break;
+                    }
+                    cap_stdout.extend_from_slice(&buf[..i]);
+                    if inner.nocapture {
+                        print!("{}", str::from_utf8(&buf).unwrap());
+                    }
+                }
+            }
+
+            if pollfds[0].revents & POLLHUP == POLLHUP && pollfds[1].revents & POLLHUP == POLLHUP {
+                break;
+            }
+        }
+
+        if Instant::now() >= next_warning {
+            let running_for = ((Instant::now() - start).as_secs() / TIMEOUT) * TIMEOUT;
+            if inner.test_threads == 1 {
+                eprint!("running for over {} seconds... ", running_for);
+            } else {
+                eprintln!(
+                    "\nlang_tests::{} ... has been running for over {} seconds",
+                    test_fname, running_for
+                );
+            }
+            last_warning = next_warning;
+            next_warning = last_warning
+                .checked_add(Duration::from_secs(TIMEOUT))
+                .unwrap();
+        }
     }
-    let output = child.wait_with_output().unwrap();
-    (output.status, output.stderr, output.stdout)
+
+    let status = child
+        .wait()
```

Comment:
Should we `wait_timeout` here, in case the test doesn't terminate for whatever reason?

Reply:
Good idea!